runtime/pkg/rilltime/rilltime.go

```diff
@@ -0,0 +1,372 @@
+}
+
+type ResolverContext struct {
+	Now        time.Time
+	MinTime    time.Time
+	MaxTime    time.Time
+	FirstDay   int
+	FirstMonth int
+}
+
+func Parse(from string) (*RillTime, error) {
+	var rt *RillTime
+	var err error
+
+	rt, err = ParseISO(from, false)
+	if err != nil {
+		return nil, err
+	}
+
+	if rt == nil {
+		rt, err = rillTimeParser.ParseString("", from)
+		if err != nil {
+			return nil, err
+		}
+		rt.IsNewFormat = true
+	}
+
+	rt.timeZone = time.UTC
+	if rt.Modifiers != nil {
+		if rt.Modifiers.Grain != nil {
+			rt.grain = grainMap[rt.Modifiers.Grain.Grain]
+			// TODO: non-1 grains
+		} else if rt.Modifiers.CompleteGrain != nil {
+			rt.grain = grainMap[rt.Modifiers.CompleteGrain.Grain]
+			// TODO: non-1 grains
+			rt.isComplete = true
+		}
+	}
+	if rt.AtModifiers != nil && rt.AtModifiers.TimeZone != nil {
+		var err error
+		rt.timeZone, err = time.LoadLocation(strings.Trim(*rt.AtModifiers.TimeZone, "{}"))
+		if err != nil {
+			return nil, fmt.Errorf("invalid time zone %q: %w", *rt.AtModifiers.TimeZone, err)
+		}
+	}
+
+	if rt.End == nil {
+		rt.End = &TimeAnchor{
+			Now: true,
+		}
+	}
+
+	return rt, nil
+}
+
+func ParseISO(from string, strict bool) (*RillTime, error) {
+	// Try parsing for "inf"
+	if infPattern.MatchString(from) {
+		return &RillTime{
+			Start: &TimeAnchor{Earliest: true},
+			End:   &TimeAnchor{Latest: true},
+		}, nil
+	}
+
+	if strings.HasPrefix(from, "rill-") {
+		// We are using "rill-" as a prefix to DAX notation so that it doesn't interfere with ISO8601 standard.
+		// Pulled from https://www.daxpatterns.com/standard-time-related-calculations/
+		rillDur := strings.Replace(from, "rill-", "", 1)
+		if t, ok := daxNotations[rillDur]; ok {
+			return Parse(t)
+		}
+	}
+
+	// Parse as a regular ISO8601 duration
+	if !durationPattern.MatchString(from) {
+		if !strict {
+			return nil, nil
+		}
+		return nil, fmt.Errorf("string %q is not a valid ISO 8601 duration", from)
+	}
+
+	rt := &RillTime{
+		Start: &TimeAnchor{Grain: &Grain{}},
+		End:   &TimeAnchor{Now: true},
+	}
+	// TODO: we do not need name based matching here since we just map the grain
+	match := durationPattern.FindStringSubmatch(from)
+	for i, name := range durationPattern.SubexpNames() {
+		part := match[i]
+		if i == 0 || name == "" || part == "" {
+			continue
+		}
+
+		val, err := strconv.Atoi(part)
+		if err != nil {
+			return nil, err
+		}
+		rt.Start.Grain.Num = &val
+		switch name {
+		case "year":
+			rt.Start.Grain.Grain = "Y"
+		case "month":
+			rt.Start.Grain.Grain = "M"
+		case "week":
+			rt.Start.Grain.Grain = "W"
+		case "day":
+			rt.Start.Grain.Grain = "d"
+		case "hour":
+			rt.Start.Grain.Grain = "h"
+		case "minute":
+			rt.Start.Grain.Grain = "m"
+		case "second":
+			rt.Start.Grain.Grain = "s"
+		default:
+			return nil, fmt.Errorf("unexpected field %q in duration", name)
+		}
+	}
+
+	return rt, nil
+}
+
+func (r *RillTime) Resolve(resolverCtx ResolverContext) (time.Time, time.Time, error) {
```

Comment:
This is a matter of convention, but we generally use "resolve" for something that makes database queries and "context" for `context.Context` objects. Could we call this `Eval(opts EvalOptions)` or something like that instead?

Reply:
Changed. So `resolveTimeRange` was called `resolveXX` just because it loaded the watermark from database?

Reply:
Yeah